File: c++/math/number-theory.cpp
// Divisores, primalidade, MDC/MMC de inteiros positivos moderados.
// Entrada: a b positivos. Ex.: 12 18 -> 1 2 3 4 6 12 / false / 6 36.
// Divisores/primo O(sqrt(a)); memória O(D) divisores e O(1) primo.
// MMC sinaliza overflow em vez de multiplicar fora do limite.
#include <iostream>
#include <limits>
#include <numeric>
#include <vector>
std::vector<long long> divisors(long long n) {
    std::vector<long long> small, large;
    for (long long d = 1; d <= n/d; ++d) if (n%d == 0) {
        small.push_back(d);
        if (d != n/d) large.push_back(n/d);
    }
    small.insert(small.end(), large.rbegin(), large.rend());
    return small;
}
bool is_prime(long long n) {
    if (n < 2) return false;
    if (n%2 == 0) return n == 2;
    for (long long d = 3; d <= n/d; d += 2) if (n%d == 0) return false;
    return true;
}
int main() {
    long long a,b; std::cin >> a >> b;
    if (a <= 0 || b <= 0) return 1;
    for (auto d : divisors(a)) std::cout << d << ' ';
    std::cout << '\n' << std::boolalpha << is_prime(a) << '\n';
    long long g = std::gcd(a,b), reduced = a/g;
    std::cout << g << ' ';
    if (reduced > std::numeric_limits<long long>::max()/b) std::cout << "overflow";
    else std::cout << reduced*b;
    std::cout << '\n';
}

File: python/math/number_theory.py
"""Divisores/primalidade para N moderado: testar até isqrt(N).
Tempo O(sqrt(N)); divisors usa O(D) memória, is_prime O(1) auxiliar.
MDC/MMC seguem Euclides; inteiros grandes têm custos maiores por operação.
Entrada: dois inteiros positivos a b.
Ex.: 12 18. Saída: 1 2 3 4 6 12 / False / 6 36.
"""
from math import gcd, isqrt
import sys


def divisors(n):
    if n < 1:
        raise ValueError("n deve ser positivo")
    small, large = [], []
    for d in range(1, isqrt(n) + 1):
        if n % d == 0:
            small.append(d)
            if d * d != n:  # raiz não aparece duas vezes
                large.append(n // d)
    return small + large[::-1]


def is_prime(n):
    if n < 2:
        return False
    if n % 2 == 0:
        return n == 2
    return all(n % d for d in range(3, isqrt(n) + 1, 2))


def lcm(a, b):
    # Alternativa a math.lcm, disponível desde Python 3.9.
    return abs(a // gcd(a, b) * b) if a and b else 0


def main():
    a, b = map(int, sys.stdin.read().split())
    print(*divisors(a))
    print(is_prime(a))
    print(gcd(a, b), lcm(a, b))


if __name__ == "__main__":
    main()
